src/java.base/share/classes/java/lang/Class.java

```diff
@@ -4148,4 +4154,28 @@ int getClassFileVersion() {
      int getClassFileAccessFlags() {
          return classFileAccessFlags;
      }
+
+    // Checks whether the class name exceeds the maximum allowed length.
+    private static boolean classNameLengthIsValid(String name) {
+        Objects.requireNonNull(name);
+        // Quick approximation: each char can be at most 3 bytes in Modified UTF-8.
+        // If the string is short enough, it definitely fits.
+        int nameLen = name.length();
+        if (nameLen <= JAVA_CLASSNAME_MAX_LEN / 3) {
+            return true;
+        }
+        // Check exact Modified UTF-8 length.
+        // The check utfLen >= nameLen ensures we don't incorrectly return true in case of int overflow.
+        int utfLen = ModifiedUtf.utfLen(name, 0);
+        return utfLen <= JAVA_CLASSNAME_MAX_LEN && utfLen >= nameLen;
```

Comment:
I would probably use early-return for the overflow case, sth like the following, to separate the normal logic from error-handling logic.

```java
if (utfLen < nameLen) {
  // overflowing...
  return false;
}

return utfLen <= JAVA_CLASSNAME_MAX_LEN;
```